### ISSUE TITLE
Add election districts

### ISSUE THREAD

**Issue by ouzor** [rOpenGov/geofi#12]
Update then data(aluejakokartat) in helsinki.

**Comment by jlehtoma** [comment]
Issue needs a bit more description.

**Comment by antagomir** [comment]
The shape files for each election district division mentioned above are available from different sources. It would be handy to have a function to retrieve the selected division, for instance: "get election districts used in Turku 2014". Helsinkin districts are available through HRI, Turku districts from Lounaispaikka

**Comment by ouzor** [comment]
What I had in mind here was function get_election_district(city, year), and then each available combination would be hard coded. So currently combinations (Helsinki, 2011), (Helsinki, 2014), and (Turku, 2014) would work, others would result in NULL.

**Comment by jlehtoma** [comment]
Is there anything else but Helsinki in the package currently (a83a0cbcb5dad598d1d7902d57bd8db5de34cf39)?

**Comment by ouzor** [comment]
No, there's currently only election maps for Helsinki region (year 2011), implemented in get_helsinki_aluejakokartat(). So it should be removed from there, and created to a new function get_election_districts(). Then (Helsinki, 2014) and (Turku, 2014) could be added to the new function.

**Comment by antagomir** [comment]
At least not for Turku but the shape files for that would be readily available from Lounaispaikka as soon as there is time to add to package. Check http://paikkatietokeskus.lounaispaikka.fi/fi/aineistot/ for "Turun Kaupungin Äänestysaluerajat" and also many other GIS data sets

**Comment by jlehtoma** [comment]
Right, I'll see what I can do for rOpenGov/gisfin#11 first and then can come back to this unless someone wants to take a stab at it.

**Comment by jlehtoma** [comment]
I created a new branch [fix/issue9](https://github.com/rOpenGov/gisfin/tree/fix/issue9/R) which has the following new features:
- Data downloading and extraction (from zip) refactored to a new function [`download_data()`](https://github.com/rOpenGov/gisfin/blob/fix/issue9/R/download_data.R)
  - Not all downloadables are necessarily zips, function tries to figure out if the target is a zip and extracts the zip
  - Implements a simple cache with overwrite capability
- `get_helsinki_aluejako()` renamed to [`get_helsinki_adminboundaries()`](https://github.com/rOpenGov/gisfin/blob/fix/issue9/R/helsinki_spatial.R) which now uses `download_data()`
  - all spatial data objects are returned in EUREF-FIN CRS (EPSG:3067)
- Added new file [`turku_spatial.R`](https://github.com/rOpenGov/gisfin/blob/fix/issue9/R/turku_spatial.R) which a single function `get_turku_adminboundaries()` which functions analogously to its Helsinki-counterpart
  - Currently 2 datasets are included: election districts and major districts (suuralueet)

**Comment by jlehtoma** [comment]
Next steps from here:
- Create a separate `get_election_districts()` which could call `get_helsinki_adminboundaries()` and `get_turku_adminboundaries()` accordingly
- Add year specifiers somewhere, currently there are none

Note also that  `get_helsinki_adminboundaries()` and `get_turku_adminboundaries()` are structurally very similar and could be refactored into a single function. The most flexible approach would be to separate the data source definitions from the implementation, e.g. a JSON or YAML file could include the necessary URLs, layer names, EPSG codes, years etc. Might be a bit of a pain to implement this, but the outcome would be much easier to maintain than coding an almost identical function for each city or creating one massive function with all the information hard coded. Thoughts?

**Comment by ouzor** [comment]
Sounds excellent! I agree that my first implementation had a lot of repetition and hard coding, so you suggestion sounds much more easier to maintain in the future when the number of data sets (and cities providing data) will grow.

**Comment by antagomir** [comment]
I propose that this is moved to [helsinki](github.com/rOpenGov/helsinki) or [election](https://github.com/rOpenGov/election) package. Or: we can add wrappers later. I group this now to Phantasy milestone.
